guibot/target.py

```diff
@@ -512,35 +512,66 @@ def load(self, steps_filename, **kwargs):
         :raises: :py:class:`errors.UnsupportedBackendError` if a chain step is of unknown type
         :raises: :py:class:`IOError` if an chain step line cannot be parsed
         """
-        if not os.path.exists(steps_filename):
-            steps_filename = FileResolver().search(steps_filename)
+        # make sure we have the correct file
+        steps_filename = self.resolve_stepsfile(steps_filename)
+        stepsfiles_seen = [steps_filename]
 
         with open(steps_filename) as f:
-            for step in f:
-                dataconfig = re.split(r'\t+', step.rstrip('\t\n'))
-                if len(dataconfig) != 2:
-                    raise IOError("Invalid chain step line '%s'" % dataconfig[0])
-                data, config = re.split(r'\t+', step.rstrip('\t\n'))
-
-                super(Chain, self).load(config)
-                self.use_own_settings = False
-
-                step_backend = self.match_settings.params["find"]["backend"]
-                if step_backend in ["autopy", "contour", "template", "feature", "tempfeat"]:
-                    data_and_config = Image(data, match_settings=self.match_settings)
-                elif step_backend in ["cascade", "deep"]:
-                    data_and_config = Pattern(data, match_settings=self.match_settings)
-                elif step_backend == "text":
-                    data_and_config = Text(data, match_settings=self.match_settings)
-                else:
-                    # in particular, we cannot have a chain within the chain since it is not useful
-                    raise UnsupportedBackendError("No target step type for '%s' backend" % step_backend)
-
-                self._steps.append(data_and_config)
+            lines = f.readlines()
+
+        while lines:
+            step = lines.pop(0)
+            dataconfig = re.split(r'\t+', step.rstrip('\t\n'))
+
+            # read a nested steps file and append to this chain
+            if dataconfig[0].endswith(".steps"):
+                nested_steps_filename = self.resolve_stepsfile(dataconfig[0])
+                # avoid infinite loops
+                if not nested_steps_filename in stepsfiles_seen:
+                    stepsfiles_seen.append(nested_steps_filename)
+                    with open(nested_steps_filename) as f:
+                        lines = f.readlines() + lines
+                continue
```

Comment:
So if the steps file has already been read, reading another line with it will result in a noop? What could be viable use cases where people would define multiple identical steps file lines within a root steps file?

Reply:
> So if the steps file has already been read, reading another line with it will result in a noop?
yes, it will be ignored

> What could be viable use cases where people would define multiple identical steps file lines within a root steps file?
No idea, but this can lead to cyclic dependencies or infinite loops

Reply:
True but we don't have to expand on all of them as reading the identical lines (with identical data and configuration) doesn't make sense for chain matching. We could thus simply raise a ValueError or a custom error if we detect repeated lines. This is more or less a fast fail to avoid something the the user has missed.

Reply:
It's an alternative approach. What I had in mind was cases when a file references another with other nested references and somewhere down the line there was a repeated stepsfile. So we can either ignore this file and carry on or fail right away, it depends on what one wants.

Reply:
I see, alright this makes sense then. I was wondering if we should put a warning but I guess this could end up in completely intended scenarios so never mind.